autonomy_core/map_plan/mapper/src/test_voxel_mapper.cpp

```diff
@@ -0,0 +1,428 @@
+#include <iostream>
+#include <gtest/gtest.h>
+#include <memory>
+#include <cmath>
+#include <chrono>
+#include <planning_ros_msgs/VoxelMap.h>
+#include "mapper/voxel_mapper.h"
+
+
+// Test class needs to be in the same namespace as VoxelMapper to be able to make it a friend class
+namespace mapper
+{
+
+/*** Declaration of fixture class to use the same configuration across tests ***/
+class VoxelMapperTest : public testing::Test
+{
+protected:
+    // Google test needs to create objects of this test fixture and it calls the default constructor
+    VoxelMapperTest();
+    VoxelMapperTest(double x_orig, double y_orig, double z_orig, int x_dim, int y_dim, int z_dim,
+                    double res, int decay_times);
+
+    const double x_origin_; 
+    const double y_origin_; 
+    const double z_origin_; 
+    const int x_dim_;
+    const int y_dim_;
+    const int z_dim_;
+    const double resolution_;
+    const int decay_times_;
+
+    // This is the VoxelMapper object that is used for all the tests
+    std::unique_ptr<mapper::VoxelMapper> p_test_mapper_;
+
+    // Timing variables to measure method callback performance
+    std::chrono::time_point<std::chrono::steady_clock> time_start, time_end;
+};
+
+
+/*** Defining the methods in the VoxelMapperTest class defintion ***/
+VoxelMapperTest::VoxelMapperTest()
+    :VoxelMapperTest(-100, -100, -5, 200, 200, 10, 0.5, 30) {}
+
+VoxelMapperTest::VoxelMapperTest(double x_orig, double y_orig, double z_orig,
```

Comment:
I don't see why you need this ctor. Ideally there should be no user defined ctor in a test class, you do all the setup in setup().

Reply:
In regards to the main question about needing three friend classes; I think the more correct way would have been to use the FRIEND_TEST macro, but I just manually added the test instead. Each test needs to be included as a friend besides the test class itself, I think the reason is because each test is a derived class from original class, the name for each derived class is a concatenation of the test class name and the test name. For example:
Code is currently:
```
class VoxelMapper {
friend class VoxelMapperTest;
friend class VoxelMapperTest_TestDecayLocalCloud_Test;
// other declarations
};
```
Code should be:
```
#include <gtest/gtest_prod.h>
class VoxelMapper {
friend class VoxelMapperTest;
FRIEND_TEST(VoxelMapperTest, TestDecayLocalCloud);
// other declarations
};
```
I'm making the change now. [Reference](https://google.github.io/googletest/advanced.html#testing-private-code)

As for the constructor, I didn't have a reason to use the constructor other than that googletest FAQ said it was usually preferred over SetUp(); mentioned [here](https://google.github.io/googletest/faq.html#CtorVsSetUp). I can make the change now.

Reply:
Great job Orlando! I especially like your clear and detailed explanation of everything you've done.

Reply:
Existing changes look great to me. Let me know when you're done with your new changes and I will test and merge!